src/Gaffer/MetadataAlgo.cpp

```diff
@@ -230,6 +231,19 @@ void setNumericBookmark( ScriptNode *scriptNode, int bookmark, Node *node )
 	IECore::InternedString metadataName( numericBookmarkMetadataName( bookmark ) );
 	for( Node *nodeWithMetadata : Metadata::nodesWithMetadata( scriptNode, metadataName, /* instanceOnly = */ true ) )
 	{
+		// Numeric bookmarks brought in by References will be ignored
```

Comment:
I don't think joining the two checks is gonna work. 

- When in a Reference, we always need to block assigning numeric bookmarks. Not just when deserialising.
- We need to block overriding numeric bookmarks outside of References via `isExecuting()`

I think what I've done now should take care of both of these and I've squashed it all down into just the one commit. Maybe have another look at it? :) Thanks!

Reply:
I wasn't trying to suggest joining the "inside reference" and the "clashing during load" checks, just refining the reference check and leaving the "clashing during load" check where it was.

I was deliberately keeping the possibility of assigning non-serialisable bookmarks temporarily inside a Reference because I thought it might be useful when navigating big graphs, but maybe that was a bad idea. If we want to enforce the can't-bookmark-inside-a-reference behaviour then our convention is to do that via the UI rather than have the API silently fail :

https://github.com/GafferHQ/gaffer/blob/master/include/Gaffer/MetadataAlgo.h#L57-L81
https://github.com/GafferHQ/gaffer/blob/master/python/GafferUI/GraphBookmarksUI.py#L59

Reply:
Sorry, I misunderstood.

I've just pushed the latest numeric bookmark (NB) code. Hopefully this is ok now.

- added the check as you suggested outside of the loop -> never loads NBs in References.
- kept the check inside the loop -> won't add NB when NB exists in graph already.
- disallowed adding NB in References through the UI as per the file you pointed me to.
- removed test that checks that the API silently fails when adding NB within Reference.

I've squashed it all into the one commit.